### ISSUE TITLE
scala.reflect -> scala.meta tree converter

### ISSUE THREAD

**Issue by xeno-by** [scalameta/scalameta#148]
The [reflect -> meta AST converter](https://github.com/scalameta/scalahost/blob/689e5b53cfc86e4b98300784e578984bc5c54beb/interface/src/main/scala/scala/meta/internal/hosts/scalac/converters/ToMtree.scala) was one of the initial pieces of functionality implemented in scalahost (the first one was the prototype for lightweight macros and the second one was the converter).

At the time when the AST converter was being developed, we had a different vision for scala.meta, which led to several fundamental assumptions that we made when developing scalahost: 1) the only environment the converter is going to be invoked is a post-typer compiler plugin to scalac, 2) it is possible to undo the desugarings that scalac does in scala.reflect trees, 3) it is practical to do so.

At the moment, one year later, we live in a different world: 1) it turned out that converting unattributed trees is also important (e.g. to support macro annotations), 2) most of desugarings [can indeed be undone](https://github.com/scalameta/scalahost/blob/f1a17f613e18673b570d70548837399deada7b84/foundation/src/main/scala/org/scalameta/reflection/Ensugar.scala), but 3) that requires [patching the typechecker](https://github.com/scalameta/scalahost/blob/689e5b53cfc86e4b98300784e578984bc5c54beb/plugin/src/main/scala/scala/meta/internal/hosts/scalac/typechecker/Analyzer.scala) to store out-of-band information about trees, and that has absolutely no chances to scale to TASTY.

Finally, the converter ended up being too smart for its own good. I wanted to experiment with a typeclass-based approach, when a macro would analyze the code of the converter and automatically infer fundeps between input and output types, but that ended up being an abomination to write and support, and now I just want it gone.

Therefore, we are planning to rewrite the converter from scratch according to the following scheme:
1. Figure out source code of the output scala.meta tree. If the input scala.reflect tree has associated source code, take it. Otherwise, prettyprint the input to obtain synthetic source code.
2. Parse the source code with the scala.meta parser to obtain the skeleton of the output. If there was actual, human-written source code, our parser will retain all syntactic details, otherwise it doesn't matter anyway.
3. Traverse the skeleton and the input in parallel, taking note of: a) symbols, b) types, c) desugarings. This conflates the ensugaring and conversion steps of the current converter, which is not that good, but, more importantly, obviates the need to guess tree shapes from scala.reflect's lossy representation, which is priceless.

This will realize our long-term dream of perfect trees that combine syntactic comprehensiveness and semantic richness, being the cornerstone of the 0.1 release: https://github.com/scalameta/scalameta/milestones.

Upd. One more year later, the vision has changed again. See https://github.com/scalameta/scalameta/issues/148#issuecomment-226984597 for more information.

**Comment by xeno-by** [comment]
In before merging the new converter, which is at the moment significantly lagging behind the old one:
  * The most recent version of Ensugar.scala: https://github.com/scalameta/scalahost/blob/689e5b53cfc86e4b98300784e578984bc5c54beb/foundation/src/main/scala/org/scalameta/reflection/Ensugar.scala
  * The most recent version of ToMtree.scala: https://github.com/scalameta/scalahost/blob/689e5b53cfc86e4b98300784e578984bc5c54beb/interface/src/main/scala/scala/meta/internal/hosts/scalac/converters/ToMtree.scala

**Comment by utaal** [comment]
I may have found a bug in the scalahost 0.0.3 compiler plugin and was about to open an issue but I noticed that this is work in progress. I tried compiling (the entirety of) github.com/non/cats with "-Ybackend:GenBCode" and it failed with `(macrosJS/compile:compileIncremental) 
scala.meta.internal.hosts.scalac.converters.ConvertException: unexpected tree during scala.reflect -> scala.meta conversion` in a subproject containing macros. Should I open an issue or wait for this rewrite to land? Here's the [full stack trace](https://gist.github.com/utaal/eaff2552cb650eaa1f04) if you're interested.

**Comment by xeno-by** [comment]
See :point_up: [September 30, 2015 1:09 PM](https://gitter.im/scalameta/scalameta?at=560bc2dcdfb3151302857d0b) for a discussion

**Comment by xeno-by** [comment]
Following the discussion at ScalaSphere (part of it documented in https://www.youtube.com/watch?v=zuBAOktT938), I'm putting this on hold.

Dealing with desugarings in scalac turns out to be very hard, and dotty is not going to improve the situation (with https://github.com/lampepfl/dotty/commit/ad483d8c3c9dc70cad9f463efd025bab7e07b882 being the current official way of going to/from AST/CST). I've already spent more than half a year on this problem, so lack of immediately useful results is my cue to focus on something else for the time being.

At the moment, I'm going to release 0.1 with what people really use (tokens, parser, our attempt at CST), cutting out everything that isn't ready yet. After writing and defending my thesis, I may revisit this issue with a fresh mind.

**Comment by xeno-by** [comment]
During the 2+ years history of scala.meta, we had two different converters: 1) an almost decent resugaring converter that worked on a lot of community projects, but didn't work on untyped trees, 2) a half-finished resugaring converter that worked on untyped and typed trees alike.

As described in the comment above, the concept of a resugaring converter has been put on hold after ScalaSphere. However, this should not prevent us from implementing a dumb converter that treats desugarings as-is.

**Comment by xeno-by** [comment]
During fall 2016, @olafurpg and I have been working on a dumb converter for scala.reflect trees. At LAMP EPFL, @liufengyun has been hacking a dumb converter to support macro annotations for Dotty. 

We all arrived at the same conclusion. Even the simplest possible converter is quite hard to implement and maintain. I think that pursuing this direction further will lead us to an engineering dead end, so I'm closing this issue for the time being.

There is hope though. See https://github.com/scalameta/scalameta/issues/604 for more details.
